### ISSUE TITLE
Keys and values of nested `with-continuation-mark` calls are evaluated before the continuation marks are updated

### ISSUE THREAD

**Issue by AlgebraicWolf** [cisco/ChezScheme#885]
It seems that currently when several nested `with-continuation-mark` calls are being applied to the same continuation, keys and values are evaluated before any continuation marks are set. This causes trouble if, for any reason, the evaluation of some key or value depends on a mark being set in the same continuation.

# Steps to reproduce

Try evaluating the following expression:

```chez
(with-continuation-mark 'a 42
    (with-continuation-mark 'b (continuation-marks-first (current-continuation-marks) 'a)
      (continuation-marks-first (current-continuation-marks) 'b)))
```

# Expected behaviour

The outermost `with-continuation-mark` set the the mark `'a -> 42`. After that, the evaluation of arguments for the innermost `with-continuation-mark` occurs. There is nothing to evaluate in `'b`, so the `(continuation-marks-first (current-continuation-marks) 'a)` gets evaluated to `42`, since this is the value currently associated with the key `'a`, and the mark `'b -> 42' is set. Finally, the expression ` (continuation-marks-first (current-continuation-marks) 'b)` is evaluated to `42`.

# Observed behaviour

The expression evaluates to `#f`.

Moreover, if the nested `with-continuation-mark` call is evaluated in the separated continuation, the behaviour matches the expected one. For example, the following expression:

```chez
(with-continuation-mark 'a 42
  (values
    (with-continuation-mark 'b (continuation-marks-first (current-continuation-marks) 'a)
      (continuation-marks-first (current-continuation-marks) 'b))))
``` 

is evaluated to `42`. If one tries to set the value in other continuation, prior to making nested `with-continuation-mark` calls, the older value is obtained. For example,

```chez
(with-continuation-mark 'a 0
  (values
    (with-continuation-mark 'a 42
      (with-continuation-mark 'b (continuation-marks-first (current-continuation-marks) 'a)
        (continuation-marks-first (current-continuation-marks) 'b)))))
```

gets evaluated to `0` instead of `42`. This highlights that the problem is not with the continuation marks in general, but with the evaluation of arguments to `with-continuation-mark`.

**Comment by mflatt** [comment]
Thanks for the report!

I think the problem is in the way the `with-continuation-mark` macro expands to `$call-consuming-continuation-attachment` and `$call-setting-continuation-attachment`:

```
(define-syntax with-continuation-mark
  (lambda (x)
    (syntax-case x ()
      [(_ key val body)
       #'($call-consuming-continuation-attachment
          '()
          (lambda (marks)
            ($call-setting-continuation-attachment
             ($update-mark marks key val)
             (lambda ()
               body))))])))
```

That consumes the current attachment before evaluating `key` and `val`, either of which might try to get marks via the attachment. The macro should expand to evaluate `key` and `val` first:

```
(define-syntax with-continuation-mark
  (lambda (x)
    (syntax-case x ()
      [(_ key val body)
       #'(let ([k key]
               [v val])
           ($call-consuming-continuation-attachment
            '()
            (lambda (marks)
              ($call-setting-continuation-attachment
               ($update-mark marks k v)
               (lambda ()
                 body)))))])))
```

**Comment by AlgebraicWolf** [comment]
@mflatt, thank you for the explanation, and for the prompt repair!

Closing this as fixed via #886.
